csep/core/brier_evaluations.py

```diff
@@ -0,0 +1,156 @@
+import numpy
+import scipy.stats
+import scipy.spatial
+
+from csep.models import EvaluationResult
+from csep.core.exceptions import CSEPCatalogException
+
+
+def bier_score_ndarray(forecast, observations):
+    """ Computes the brier score
+    """
+
+    observations = (observations >= 1).astype(int)
+    prob_success = 1 - scipy.stats.poisson.cdf(0, forecast)
+    brier = []
+
+    for p, o in zip(prob_success.ravel(), observations.ravel()):
+
+        if o == 0:
+            brier.append(-2 * p ** 2)
+        else:
+            brier.append(-2 * (p - 1) ** 2)
+    brier = numpy.sum(brier)
+
+    for n_dim in observations.shape:
+        brier /= n_dim
+
+    return brier
+
+
+def _simulate_catalog(sim_cells, sampling_weights, sim_fore, random_numbers=None):
+    # Modified this code to generate simulations in a way that every cell gets one earthquake
+    # Generate uniformly distributed random numbers in [0,1), this
+    if random_numbers is None:
+        # Reset simulation array to zero, but don't reallocate
+        sim_fore.fill(0)
+        num_active_cells = 0
+        while num_active_cells < sim_cells:
+            random_num = numpy.random.uniform(0,1)
+            loc = numpy.searchsorted(sampling_weights, random_num, side='right')
+            if sim_fore[loc] == 0:
+                sim_fore[loc] = 1
+                num_active_cells = num_active_cells + 1
+    else:
+        # Find insertion points using binary search inserting to satisfy a[i-1] <= v < a[i]
+        pnts = numpy.searchsorted(sampling_weights, random_numbers, side='right')
+        # Create simulated catalog by adding to the original locations
+        numpy.add.at(sim_fore, pnts, 1)
+
+    assert sim_fore.sum() == sim_cells, "simulated the wrong number of events!"
+    return sim_fore
+
+
+def _brier_score_test(forecast_data, observed_data, num_simulations=1000, random_numbers=None,
+                            seed=None, use_observed_counts=True, verbose=True):
+    """  Computes binary conditional-likelihood test from CSEP using an efficient simulation based approach.
+
+    Args:
+
+    """
+
+    # Array-masking that avoids log singularities:
+    forecast_data = numpy.ma.masked_where(forecast_data <= 0.0, forecast_data)
```

Comment:
Beware of the potential side effects of filtering away zero-rate bins, see #226 (tl;dr: allows to cheat).

Reply:
How about negative rates (just in case)?

Reply:
The rates should be always positive. If we estimate them from catalogue-based forecasts as the average (per synthetic catalog) number of points per bin we should never get negative rates. If the rates come from a grid-based forecast then it should be an indication that there is a problem with the model on how these rates are generated. More specifically, the Brier score does not work directly with the rates, it works by estimating the probability that a bin is active (percentage of synthetic catalogues with at least 1 event in that bin), so zero-rates bins have zero probability, but bins with the same (positive) rate may have different probabilities depending on the variance. The rate completely determines the probability only in the case of grid-based forecasts assuming an independent Poisson distribution for the number of events in each bin.  In both cases, (I think) it should be impossible to get negative rates, if the forecast is catalogue-based this is by construction, if the forecast is grid-based with Poisson counts then the rate should always be greater or equal to zero by definition of the Poisson distribution. The same holds if we consider a Negative Binomial distribution which has both parameters, and the mean, greater or equal zero.

Reply:
The first sentence should be "The rates should be always non-negative"

Reply:
Thanks for the quick explainer, Francesco!

This reminds me that we should clarify that this PR is currently for grid-based forecasts only.

I only ask about negative rates to think about necessary "safety" measures (currently there are none, apart the silent masking here and in `binomial_evaluations`). Since ≤ 0 rates could in principle be used for cheating in LL-based tests when masking them away (#226), I wanted to inquire whether similar things are possible with the Brier score.

> The first sentence should be "The rates should be always non-negative"

Ah, sure; I mentally replaced positive with non-negative when I read your first sentence ;-) (Btw: you can always edit your posts)

Reply:
I edited and removed the last comment :) Anyway, I think that a possible solution could be to impose them to zero and rise a warning saying that some of the rates were negative. Excluding them may be used to cheat by placing them in bins that a malicious user wants to exclude from the computations.

Reply:
Yes, spitting out at least a Warning in those cases seems a common desire in #226. But there are good reasons to _not_ "touch"/"fix" the original forecasts - at least the zero rates. In any case, we should continue this discussion there.